player_linux.go

```diff
@@ -0,0 +1,126 @@
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License
+// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+// or implied. See the License for the specific language governing permissions and limitations under
+// the License.
+
+package oto
+
+// #cgo LDFLAGS: -lasound
+// #include <alsa/asoundlib.h>
+// #include "player_linux.h"
+import "C"
+
+import (
+	"fmt"
+	"unsafe"
+)
+
+type player struct {
+	handle     *C.snd_pcm_t
+	buf        []byte
+	bufSamples int
+}
+
+func alsaError(err C.int) error {
+	return fmt.Errorf("oto: ALSA error: %s", C.GoString(C.snd_strerror(err)))
+}
+
+func newPlayer(sampleRate, numChans, bytesPerSample, bufferSizeInBytes int) (*player, error) {
+	var p player
+
+	// open a default ALSA audio device for blocking stream playback
+	errCode := C.snd_pcm_open(&p.handle, C.CString("default"), C.SND_PCM_STREAM_PLAYBACK, 0)
+	if errCode < 0 {
+		return nil, alsaError(errCode)
+	}
+
+	// bufferSize is the total size of the main circular buffer fullness of this buffer
+	// oscilates somewhere between bufferSize and bufferSize-periodSize
+	bufferSize := C.snd_pcm_uframes_t(bufferSizeInBytes / (numChans * bytesPerSample))
+	// periodSize is the number of samples that will be taken from the main circular
+	// buffer at once, we leave this value to bufferSize, because ALSA will change that
+	// to the maximum viable number, obviously lower than bufferSize
+	periodSize := bufferSize
+
+	// choose the correct sample format according to bytesPerSamples
+	var format C.snd_pcm_format_t
+	switch bytesPerSample {
+	case 1:
+		format = C.SND_PCM_FORMAT_S8
+	case 2:
+		format = C.SND_PCM_FORMAT_S16_LE
+	default:
+		panic(fmt.Errorf("oto: bytesPerSample can be 1 or 2, got %d", bytesPerSample))
+	}
+
+	// set the device hardware parameters according to sampleRate, numChans, format, bufferSize
+	// and periodSize
+	//
+	// bufferSize and periodSize are passed as pointers, because they may be changed according
+	// to the wisdom of ALSA
+	//
+	// ALSA will try too keep them as close to what was requested as possible
+	errCode = C.ALSA_hw_params(p.handle, C.uint(sampleRate), C.uint(numChans), format, &bufferSize, &periodSize)
+	if errCode < 0 {
+		p.Close()
+		return nil, alsaError(errCode)
+	}
+
+	// allocate the buffer of the size of the period, use the periodSize that we've got back
+	// from ALSA after it's wise decision
+	p.bufSamples = int(periodSize)
+	p.buf = make([]byte, 0, p.bufSamples*numChans*bytesPerSample)
+
+	return &p, nil
+}
+
+func (p *player) Write(data []byte) (n int, err error) {
+	for len(data) > 0 {
+		// cap(p.buf) is equal to the size of the period
+		toWrite := min(len(data), cap(p.buf)-len(p.buf))
+		p.buf = append(p.buf, data[:toWrite]...)
+		data = data[toWrite:]
+		n += toWrite
+
+		// our buffer is not full yet, we won't flush it yet
+		if len(p.buf) < cap(p.buf) {
+			continue
```

Comment:
Hmm, is this really true that `len(p.buf) < cap(p.buf)` is true iff `len(p.buf) == 0`? I'm now reading the code.

Let's put `break` here instead of `continue`, and this makes this logic more explicit.